src/Invoice/Calculator/FinancialYearCalculator.php

```diff
@@ -0,0 +1,114 @@
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace App\Invoice\Calculator;
+
+use App\Configuration\SystemConfiguration;
+
+class FinancialYearCalculator
+{
+    private \DateTimeInterface|null $financialYearStart = null;
+    private \DateTimeInterface|null $financialYearEnd = null;
+
+    public function __construct(private SystemConfiguration $systemConfiguration)
+    {
+        if (!$financialYearStart = $this->systemConfiguration->getFinancialYearStart()){
+            return;
+        }
+        $this->financialYearStart = \DateTime::createFromFormat('Y-m-d', $financialYearStart);
+        $this->setFinancialYearEndFromStartDate($this->financialYearStart);
+    }
+
+    private function setFinancialYearEndFromStartDate(\DateTimeInterface $financialYearStart): void
+    {
+        $this->financialYearEnd = $financialYearStart
+            ->add(\DateInterval::createFromDateString('1 year'))
+            ->sub(\DateInterval::createFromDateString('1 day'));
+    }
+
+    /**
+     * @throws FinancialYearNotSetException
+     */
+    private function isYearPrevious(\DateTimeInterface $dateTime): bool
+    {
+        if (!$this->financialYearStart){
+            throw new FinancialYearNotSetException('Financial year not set!');
+        }
+
+        $financialYearStart = clone $this->financialYearStart;
+
+        $financialYearStart->setDate(
+            $dateTime->format('Y'),
+            $financialYearStart->format('m'),
+            $financialYearStart->format('d')
+        );
+
+        return $dateTime->getTimestamp() < $financialYearStart->getTimestamp();
+    }
+
+    /**
+     * @throws FinancialYearNotSetException
+     */
+    public function getLongFinancialYear(\DateTimeInterface $dateTime): string
+    {
+        if ($this->isYearPrevious($dateTime)) {
+            return $dateTime->format('Y') - 1;
+        }
+
+        return $dateTime->format('Y');
+    }
+
+    /**
+     * @throws FinancialYearNotSetException
+     */
+    public function getShortFinancialYear(\DateTimeInterface $dateTime): string
+    {
+        if ($this->isYearPrevious($dateTime)) {
+            return $dateTime->format('y') - 1;
+        }
+
+        return $dateTime->format('y');
+    }
+
+    /**
+     * @throws FinancialYearNotSetException
+     */
+    public function getFinancialYear(\DateTimeInterface $dateTime): string
+    {
+        if ($this->isYearPrevious($dateTime)) {
+            return ($dateTime->format('Y') - 1) . '-' . $dateTime->format('y');
+        }
+
+        return $dateTime->format('Y') . '-' . ($dateTime->format('y') + 1);
+    }
+
+    /**
+     * @throws FinancialYearNotSetException
+     */
+    public function getFinancialYearStart(): \DateTimeInterface|bool
+    {
+        if (!$this->financialYearStart) {
+            throw new FinancialYearNotSetException('Financial year not set!');
```

Comment:
All other errors use `InvalidArgumentException`, e.g.
```
throw new \InvalidArgumentException('Customer has no number, replacer {cnumber} failed evaluation');
```
I think that is sufficient. Also the message doesn't need a `!` at the end 😁

Reply:
Fixed.